### PR TITLE
Reduce resharding impact by redirecting data to new shards

### PR DESCRIPTION
**Pull request by Harkishen-Singh**
Signed-off-by: Harkishen-Singh <harkishensingh@hotmail.com>

Fixes: #7230 

This PR adds support to reduce resharding impact by spinning up new shards at the time of resharding if there are existing shards. The new shards are fed with data from existing shards while blocking the incoming samples. Once this is done, the incoming samples are consumed. This keeps the samples in order. After this, the old shards are replaced with these new shards. This reduces the wait time of the samples due to slow shards, as mentioned in the related issue.

### CODE REVIEW

**Comment by Harkishen-Singh** [comment]
cc @cstyan @csmarchbanks

**Comment by csmarchbanks** [comment]
Thank you for the PR! 

We are aware of this PR and will review it when we have time, probably Monday or Tuesday. Please avoid re-pinging people after only a couple days, especially when those days are a weekend.

**Comment by Harkishen-Singh** [comment]
Ahh, sorry. I didn't aim to re-ping, I felt like mentioning in PR description is not useful, so removed from there and mentioned in the comment.

File: storage/remote/queue_manager.go
Comment (csmarchbanks) [inline storage/remote/queue_manager.go:921]: Should this close the channel instead of writing a nil marker? Same with down below, close the channel once done writing to it.

**Review by csmarchbanks — COMMENTED** [review]
No worries, we will get notified if it is in the description or a comment.

I left a few comments, I am not sure this is quite correct yet. It looks like `runShard()` would currently continue to pick up samples from its `queue` and try to send them rather than put them into a queue buffer, or a new queue. Then, only the very last request would be moved to the new queues in the current implementation?

**Comment by Harkishen-Singh** [comment]
> It looks like runShard() would currently continue to pick up samples from its queue and try to send them rather than put them into a queue buffer, or a new queue. Then, only the very last request would be moved to the new queues in the current implementation?

As soon as the queue channel is closed, the pending samples in the shard will be sent to the buffer. It doesn't matter if its the last request or not, the pending samples are sent to buffers than waiting for them to be flushed (which is responsible for a slow shard).

Do we want also want to stop the existing `sendSample` (which is in process of sending) that may be slow? (assuming the answer to be no)

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:493]: That should be better, than looping across millions of samples/series.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:326]: The reason for reusing `Append()` is that the functionality is common with respect to both the use cases. Do you want think making this separate would be better?

**Comment by csmarchbanks** [comment]
The issue is that using the default queue config, the `queue` channel might have 2,500 samples in it. All of those would be processed, and 4-5 requests would be sent, before the `close` is processed, at which point up to 500 samples would be moved over. Using those numbers, resharding would speed up by about 20%, but I think we can do better :).

I have seen queue configs where 100,000 samples might be buffered in the `queue` channel., which you can imagine may take quite a long time to successfully send to the remote before seeing that the channel was closed.

Does that make sense?

File: storage/remote/queue_manager.go
Comment (csmarchbanks) [reply storage/remote/queue_manager.go:326]: The functionality is similar, but different enough that I would prefer to keep the strong typing on the `Append` function. If some of the inner code is shared, you could refactor some of it into a third function.

**Comment by Harkishen-Singh** [comment]
@csmarchbanks as far as i understand, as soon as the a queue channel is closed, the following iteration in `runShard` will make ok as false and hence shift the pending samples to the buffer and then to the new shards. The only way it can be slow is if the queue channel is closed but at that time, `sendSamples()` inside `if nPending >= max` is running blockingly, meaning we have to wait for that send to finish, which can make it slow, and hence i asked if we want to stop the currently running `sendSamples()`.

Maybe, if you could explain more on how `All of those would be processed, and 4-5 requests would be sent` is the case. That would be very helpful.

**Comment by csmarchbanks** [comment]
Ahh, I think I might see where the misunderstanding is. When a channel is closed, `ok` only will be false when all of the samples from that channel are also drained. Since that channel can have thousands to hundreds of thousands of samples in it, all of those would be processed by the normal batch and send logic before the new logic is seen when `ok` comes back as false.

Here is a go playground example that shows the behavior in a simpler manner: https://play.golang.org/p/lmZ2po7P3OH. As can be seen, if we have 2300 samples in the queue when it is closed, we will send 4 requests of 500 samples before moving the final 300 samples to the new queue.

Did this explanation help at all?

**Comment by Harkishen-Singh** [comment]
> Did this explanation help at all?

Yes, very much. Thank you for the awesome explanation. I will update the code soon.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go]: Yes, that is better.

File: storage/remote/queue_manager.go
Comment (cstyan) [inline storage/remote/queue_manager.go:850]: the comment here seems incorrect?

File: storage/remote/queue_manager.go
Comment (cstyan) [inline storage/remote/queue_manager.go:370]: does this need to be exported?

File: storage/remote/queue_manager.go
Comment (cstyan) [inline storage/remote/queue_manager.go:871]: The name of this variable is a bit hard to follow when it's being used like this. As far as I can tell you're only using it here, and only to block enqueuing of new samples if we're resharding.

File: storage/remote/queue_manager.go
Comment (cstyan) [reply storage/remote/queue_manager.go:850]: this almost seems better to just inline in the reshard block, can't think of a name that's actually descriptive

File: storage/remote/queue_manager.go
Comment (cstyan) [inline storage/remote/queue_manager.go:792]: Having `stop` and `stopOldShards` is again confusing naming IMO. Can we can name this one something like `stopForResharding` or something to that effect?

**Review by cstyan — COMMENTED** [review]
few comments, having a bit of a hard time following the changes having come in to this late, but it seems like you and Chris are on top of things :)

File: storage/remote/queue_manager.go
Comment (csmarchbanks) [inline storage/remote/queue_manager.go:703]: Do these need to be pointers?

File: storage/remote/queue_manager.go
Comment (csmarchbanks) [inline storage/remote/queue_manager.go:931]: Why are we closing the queue here? Readers should never close the channels (it might also be worthwhile to explicitly make `queue` read only).

File: storage/remote/queue_manager.go
Comment (csmarchbanks) [inline storage/remote/queue_manager.go:925]: When would this happen, and is it guaranteed to only happen once?

**Review by csmarchbanks — COMMENTED** [review]
I left some comments, but generally I agree with Callum, I am having a hard time understanding everything happening. I think the overall workflow is mostly correct, but debugging issues will be painful and I will need more time to review fully. The following is a suggestion that might make it easier to understand :)

Rather than having a having logic spread around among several methods in a `shards` struct, would it be clearer to represent this logic by creating a brand new `shards` struct that is then populated with values from the old `shards`? Resharding would look something like:
```
newShards := t.newShards()
newShards.start(numShards)
// transferTo would stop the shard, and enqueue pending samples to `newShards`.
t.shards.transferTo(newShards)
t.shards = newShards
```
But with some additional locking to make sure `Append` is still safe. Thoughts?

If the above sounds good, I would recommend two pull requests, one to refactor `shards` to be idempotent and replaced during reshard, and a second to implement the `transferTo` logic.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:850]: Yeah, sorry, missed changing the naming of the function.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:871]: Sure, I think, renaming to `blockIncomingSamples` along with a descriptive comment should be good.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:844]: I used sorting since we wanted to ensure that samples should be sorted always in the queues. But if you feel that the will eventually be sorted during streaming, then sorting can be avoided. For now, going with not using the sorts.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:837]: Agreed. Going with sorts in that case.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:834]: I am not sure how will `wg.Done()` get missed, but using `defer` here is always good. Thank you.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:844]: We anyways need to sort, else I am not sure if sortedness is guaranteed.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:862]: Sorry, but the reason I gave it isPending is, to say in a way that we are enqueuing a pending sample (pending as they were already present in old shards). But, yes force does sound better if we go by the name of the function.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:925]: I have updated the code. This is to accept the samples that would have entered the queue, between when the resharding signal is sent to `t.shards.reshard` and blocking incoming samples at the time of creation of newQueues.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:789]: After the recent push, I feel that `stopOldShards()` is not required anymore.

**Comment by Harkishen-Singh** [comment]
With the recent code pushed, I feel that `stopOldShards()` is not required any more. This is because stopOldShards() is called only during resharding and when a reshard signal is sent, it goes into the t.reshard case, and then accepts incoming samples and redirecets to new queues, ending the old shards/go_routines. 

What do other think on this?

I will try to make the code more cleaner. Sorry for the inconveniences :sweat_smile:

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:833]: > Might be easier to just drain every buffer one at a time if we really need to sort them.

Oh yes, if I remove the go-routines here, then there should not be any issue.

**Comment by Harkishen-Singh** [comment]
> If the above sounds good, I would recommend two pull requests, one to refactor shards to be idempotent and replaced during reshard, and a second to implement the transferTo logic.

mmm, they both are part of a single feature, then why for two PRs? Also, as far as I have understood, I think for the implementation, they both are required.

**Comment by Harkishen-Singh** [comment]
@csmarchbanks I have updated with the suggestions. The renaming suggestions were helpful and I think the code looks more understandable now. There are some changes like the `enqueue()` now does not need to send false for blocking, instead is being handled by queuemanager and related changes that I will do after listening your thoughts on the updated implementation.
Thank you.

File: storage/remote/queue_manager.go
Comment (csmarchbanks) [inline storage/remote/queue_manager.go:815]: I don't think it is safe to stop reading from the queue as soon as we block new enqueues. There will still be a potential backlog of samples to work through. I think just a for loop across all samples in `queue` and adding them to pendingSamples will do the trick.

File: storage/remote/queue_manager.go
Comment (csmarchbanks) [inline storage/remote/queue_manager.go:835]: In this path where is `s.queue` closed? It doesn't look like `s.stop()` is ever called for the old shards on a reshard?

**Review by csmarchbanks — COMMENTED** [review]
Thanks for all your work, this is certainly becoming more understandable!

What would happen in the current implementation if a reshard is attempted but the remote endpoint is down? You would have shards stuck in sendSamples forever without a context cancellation I think?

Also, there are some lint failures, I would love to see the results of the tests, I think there are some bugs, and it might be that we need better test coverage if they all pass.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:835]: Yes, I think I left that. `s.queue` should definitely closed.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:815]: Yes, this is better. Thank you for pointing it.

**Comment by Harkishen-Singh** [comment]
> What would happen in the current implementation if a reshard is attempted but the remote endpoint is down? You would have shards stuck in sendSamples forever without a context cancellation I think?

For context cancellation or hard shutdown, should we use a TTL or limit a particular number of times a non-decreasing pending samples happens?

**Comment by Harkishen-Singh** [comment]
> What would happen in the current implementation if a reshard is attempted but the remote endpoint is down? You would have shards stuck in sendSamples forever without a context cancellation I think?

The current implementation (with recent push) is such that after each shard is formed, we check if send `sendSample()` fails to send any sample, which is beyond the flush deadline, then we do a hardshutdown like earlier cases.

**Comment by roidelapluie** [comment]
I am wondering it it is worth complicating this code now, since there is a consensus to make remote_write transactional, so all this code would change anyway.

https://docs.google.com/document/d/1vhXKpCNY0k2cbm0g10uM2msXoMoH8CTwrg_dyqsFUKo/edit

File: storage/remote/queue_manager.go
Comment (csmarchbanks) [inline storage/remote/queue_manager.go:920]: This means that any retries that take longer than the flushDeadline will cause a hard shutdown? Can we just use the context cancel to stop sending?

File: storage/remote/queue_manager.go
Comment (csmarchbanks) [inline storage/remote/queue_manager.go:769]: I don't really like that this resharding logic is done inside of stop. A shard stop should just worry about itself, not the queue manager.

**Review by csmarchbanks — COMMENTED** [review]
I left a few more comments.

> I am wondering it it is worth complicating this code now, since there is a consensus to make remote_write transactional, so all this code would change anyway.

Is anyone actually working on making remote write transactional? I couldn't attend the last dev summit so may have missed it. If not I think this bit of extra complexity is worthwhile until transactional rw is out.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:524]: Ah, that's the debug stuff. I forgot to remove.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:920]: > This means that any retries that take longer than the flushDeadline will cause a hard shutdown?

Yes

> Can we just use the context cancel to stop sending?

Sorry, do you mean to just skip the current send or to stop the shard? The current implementation here stops the reshard by using the context's cancel (hardShutdown is the cacncelFunc). If not, then I think I did not get your point. Can you please explain a bit more?

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go]: Yeah, this was wrong. I have shifted this to `reshardLoop()` which controls the transfer of shards, etc. This should be the place for blocking.

**Comment by csmarchbanks** [comment]
Hi @Harkishen-Singh, I am still on vacation this week, I plan to take another look at this sometime next week!

File: storage/remote/queue_manager.go
Comment (csmarchbanks) [reply storage/remote/queue_manager.go:1056]: I will rephrase, right now `(*shards).stop()` contains all of the logic around soft and hard shutdowns. It would be ideal if that logic would still only be contained in `stop()`. It is not obvious to be calling `hardShutdown` from inside a method called `sendSamples`, and I am sure that will lead to unexpected behavior in the future.

I think that ends up meaning that 
```
	if resharding {
		return
	}
```
inside of `stop()` should be removed. Since the shard would be sending to the new queue instead of to the remote, shutdown should still happen pretty quickly. You may have to setup some background processing in reshardLoop to avoid blocking on stop, but that is preferable IMO.

File: storage/remote/queue_manager.go
Comment (csmarchbanks) [inline storage/remote/queue_manager.go:795]: I think there is a possible race condition on this assignment right now. `(*QueueManager).Append()` tries to read from `t.shards` even if incoming samples are blocked right now.

File: storage/remote/queue_manager.go
Comment (csmarchbanks) [inline storage/remote/queue_manager.go:996]: I don't think you can just keep adding to `pendingSamples` like this. The slice has a size equal to the max samples per send, which is almost always smaller than the capacity the `queue` channel. I think it would be necessary either to batch up pendingSamples, or just send the samples directly without converting them to pending and back.

File: storage/remote/queue_manager.go
Comment (csmarchbanks) [inline storage/remote/queue_manager.go:793]: Shouldn't sends be blocked until the new shards replace the old shards (two lines down)?

**Review by csmarchbanks — COMMENTED** [review]
Tried to answer your question and left a few more comments. Looks like the remote tests are timing out, which is likely related to a block somewhere in the reshard code.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go]: Replaced this with a lock.

**Comment by Harkishen-Singh** [comment]
Hey @csmarchbanks. I have updated the code with a more cleaner approach in lines with the suggestions. Thank you.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:1063]: This has been addressed. I think this can be resolved now.

**Comment by codesome** [comment]
@csmarchbanks could you take a look at this PR? Thanks!

(PS: @Harkishen-Singh the tests are failing)

**Comment by csmarchbanks** [comment]
Now that you are working on transactional remote write, do you think this PR is still worthwhile? If so, can we get the tests passing before another round of reviews since there were issues last time I commented. Otherwise, let's put it on the back burner to focus on other work.

**Comment by Harkishen-Singh** [comment]
PS: This isn't ready for review yet. I think the new Github features automatically asks for reviews from the maintainers.

**Comment by roidelapluie** [comment]
Yes, unless your mark the PR as draft.

**Comment by codesome** [comment]
@csmarchbanks @cstyan can you take another look at this, please?

**Comment by codesome** [comment]
(closed by mistake)

File: storage/remote/queue_manager.go
Comment (cstyan) [inline storage/remote/queue_manager.go:908]: did you mean to change this comment?

File: storage/remote/queue_manager.go
Comment (cstyan) [inline storage/remote/queue_manager.go:929]: function name has changed

File: storage/remote/queue_manager.go
Comment (cstyan) [inline storage/remote/queue_manager.go:831]: we can just use `len(s.queues)` where you're using the new var

File: storage/remote/queue_manager.go
Comment (cstyan) [inline storage/remote/queue_manager.go:1002]: I think you could simplify this loop to:
```
// Accept those samples that might have entered the queue between sending the resharding
// signal and blocking of incoming samples at the time of creation of newer queues.
for sample :=  range queue {
    samples = append(samples, sample)
}
```

File: storage/remote/queue_manager.go
Comment (cstyan) [inline storage/remote/queue_manager.go:1012]: This could be very noisy when many shards are being run, perhaps it would be better to move this to debug level and have the overall `bufferize` function print a line at info level about the total # of samples passed to the new shards?

**Review by cstyan — COMMENTED** [review]
You'll need to rebase as well.

**Comment by Harkishen-Singh** [comment]
Rebased and updated.

**Comment by cstyan** [comment]
Looks like there's now a data race and also maybe a routine that never ends somewhere in the resharding procedure, take a look at the failing tests.

**Comment by Harkishen-Singh** [comment]
Ah sorry, I missed a lock. It should be fine now.

**Review by ALESYAkot — APPROVED** [review]

**Comment by bboreham** [comment]
Is this likely to be resurrected?  I see the original issue #7230 is still open.

**Comment by cstyan** [comment]
I personally don't have time to take this over. @Harkishen-Singh do you plan/have any interest in picking this up again or should we mark it as open for contributors?

**Comment by roidelapluie** [comment]
We have looked at this pull request during our bug scrub.

Given the lack of response, I have marked the issue as `help wanted`, and we have decided to close this PR.

Thank you for your contribution.

**Pull request closed without merging**
